### PR TITLE
JBPM-6342 Pressing [ESC] does not cancels operations.

### PR DESCRIPTION
**Pull request by tiagodolphine**
Now cancel operation pressing ESC is working. It was implemented on these points:

- palette cancel item drag to canvas, refactory on the **ShapeGlyphDragHandlerImpl** to support cancel and use the **DragProxy** interface.
- canvas element selection cancel, using **KeyboardControl**
- canvas element drag cancel, using **KeyboardControl**
- canvas connector creation cancel, using **KeyboardControl**

Unit tests were added to the new points and to some old missing points.

Obs:  I think the DMN Palette could be refactored there are common code with the BS3 Palette, but probably could be made on other JIRA.
 
@romartin 
@manstis 
@hasys

### CODE REVIEW

**Comment by hasys** [comment]
Jenkins execute full downstream build

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/glyph/ShapeGlyphDragHandlerImpl.java [hunk 1]
```diff
@@ -88,39 +123,55 @@ private void setDragProxyPosition(final LienzoPanel dragProxyPanel,
...
-                                                               final double y = mouseUpEvent.getY();
-                                                               callback.onComplete(x,
+        addMouseUpEvent(floatingPanel,
+                        callback);
+    }
+
+    private void addMouseMoveEvents(LienzoPanel floatingPanel,
+                                    DragProxyCallback callback,
+                                    Style style) {
+        handlerRegistrations.add(RootPanel.get().addDomHandler(mouseMoveEvent -> {
```
[the remark this answers is not in the document]
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/glyph/ShapeGlyphDragHandlerImpl.java]: Actually I haven't changed the behavior, if you look at the master branch it is already adding the handlers to the RootPanel. On the refactory I tried to attach to only the Palette or Canvas but this not worked, the only way it worked was with the RootPanel as it current is.
About the session.getKeyboardControl this was the first try but there are no reference to the session in this point, do you have any idea ? I've even tried @EventHander on the Palette View and not worked as well.
Anyway this palette point were the more complicated point I've spend some time testing different things but this was the more decoupled way that worked well, but open for suggestions.

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/glyph/ShapeGlyphDragHandlerImpl.java [hunk 2]
```diff
@@ -88,39 +123,55 @@ private void setDragProxyPosition(final LienzoPanel dragProxyPanel,
...
+                                                                   final int x = mouseUpEvent.getX();
+                                                                   final int y = mouseUpEvent.getY();
+                                                                   callback.onComplete(x,
+                                                                                       y);
+                                                               },
+                                                               MouseUpEvent.getType()));
+    }
+
+    private void addKeyboardEscHandler() {
+        handlerRegistrations.add(RootPanel.get().addDomHandler(k -> clear(),
```
[the remark this answers is not in the document]
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/components/glyph/ShapeGlyphDragHandlerImpl.java]: The same as I answered before, this was the my first try but in this point I didn't find a way to access the session after testing different things this was the way it worked for all cases, but open for suggestions.

**Comment by tiagodolphine** [comment]
Hi @manstis thanks for the points in fact, your concerns were mine and I really wanted to use the KeyboardControl on all the points, but the palette drag-drop to canvas was the more complicated point, the only way it worked was adding the handlers on the ShapeGlyphDragHandlerImpl in the same way it was being made with the Mouse events. 
And AFAIK the handlers are transient, so we add them on the item drag start and remove on drag end / cancel.

**Comment by hasys** [comment]
Jenkins execute full downstream build

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java [hunk 3]
```diff
@@ -74,6 +83,32 @@ public DragControlImpl(final CanvasCommandFactory<AbstractCanvasHandler> canvasC
...
+    }
+
+    private void clear() {
+        dragContext = null;
+        selectedElement = null;
+    }
+
+    @Override
+    public void unbind() {
+
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java]: Probably it's missing some action to perform here? Like de-register `this` instance from the session's keyboard control? (Not sure how this part works yet as it's quite new, let's see if @manstis can confirm).

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java [hunk 4]
```diff
@@ -64,6 +69,10 @@
 
     protected final double[] dragShapeSize = new double[]{0, 0};
 
+    protected DragContext dragContext;
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java]: FYI you're not overriding the `disable` method in this implementation, not sure if makes sense, just let you know that if these instance variables need to be clean you can override and do it in the `disable`, if necessary.

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/select/SelectionControlImpl.java [hunk 5]
```diff
@@ -60,6 +65,25 @@ public SelectionControlImpl(final Event<CanvasElementSelectedEvent> elementSelec
...
+                                    KeyboardEvent.Key.ESC)) {
+            if (Objects.nonNull(selectedElementUUID)) {
+                deselect(selectedElementUUID);
+            }
+        }
+    }
+
+    @Override
+    public void unbind() {
+
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/select/SelectionControlImpl.java]: same as for `DragControlImpl`, not sure if necessary to do some cleans here. See comment above.

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/ConnectionAcceptorControlImpl.java [hunk 6]
```diff
@@ -47,10 +52,30 @@
...
+    void onKeyDownEvent(final KeyboardEvent.Key... keys) {
+        if (KeysMatcher.doKeysMatch(keys,
+                                    KeyboardEvent.Key.ESC)) {
+            cancelCanvasActionEvent.fire(new CancelCanvasAction(getCanvasHandler().getCanvas()));
+        }
+    }
+
+    @Override
+    public void unbind() {
+
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/ConnectionAcceptorControlImpl.java]: same comment as for `DragControlImpl`

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java (hunk 3 above)
Comment (manstis) [reply kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java]: {{KeyboardControl}} doesn't have anything for "unbind" (perhaps it should!)

**Comment by hasys** [comment]
Hi @tiagodolphine, I tried Workbench from FullDownstream build [You can download it here.](https://kie-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/kie-wb-common-downstream-pullrequests/308/artifact/downstream-repos/kie-wb-distributions/kie-wb-parent/kie-wb-distribution-wars/kie-wb/target/), but Esc key is not working there. Is it already integrated to Workbench?

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java (hunk 4 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/drag/DragControlImpl.java]: @romartin implemented `disable`

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/ConnectionAcceptorControlImpl.java (hunk 6 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/ConnectionAcceptorControlImpl.java]: @romartin just inserted a comment indicating nothing to unbind for now...

**Comment by tiagodolphine** [comment]
Hi @hasys I just built and gwt:run the kie-wb and it seems to be ESC working. Does it possible the fulldownstream build on CI don't pull the latest version? 
screencast_15-09-2017

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by romartin** [comment]
@tiagodolphine it seems it's missing to fix a checkstyle violation

**Comment by manstis** [comment]
@tiagobento Please follow up this PR to completion.

**Comment by romartin** [comment]
@manstis FYI Tiago on PTO this week, so his PRs will be not able to be merged until he's back... if it's good for you

**Comment by tiagobento** [comment]
@manstis You mean me or @tiagodolphine? I assume it's a typo but thought I should ask this time :)

**Comment by manstis** [comment]
Yes sorry...

**Comment by hasys** [comment]
Previous fail was caused by my changes droolsjbpm-build-bootstrap repository, now it have to be OK.
Jenkins execute full downstream build

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Hi @hasys about the `canvas element selection cancel - works partially - element lose blue borders, but toolbox still present` this was fixed and already updated on the PR.
The `canvas element drag cancel - works partially` I made some tests and investigated how to implement that but it isn't simple, maybe in this case a new JIRA would be better, what do you think about it?

cc @romartin @manstis

**Comment by romartin** [comment]
Hey @tiagodolphine @hasys @manstis , 

IMO I just would try to go ahead and merge this PR as it is, if we detect some issues yeah let's open new tickets, but at least adding this commit makes it behave better than it was before... so all good for me.

Anyway thanks for the reviews and comments guys, good job!

**Pull request merged**